scip interface (first version)

tias opened the pull request:
with help of Mark Turner from SCIP.

Now... turns out that @IgnaceBleukx  already worked on a SCIP interface too, about a year ago (e.g. [SCIP branch](/CPMpy/cpmpy/tree/scip)

So maybe Ignace can now make a best of both worlds? : )
(current branch is up to date with current template, but the __add__ part is minimal, e.g. reified linear is not properly tested/supported at the least).

tias committed:
first minimally working version of scip interface, thanks Mark Turner for the help

IgnaceBleukx committed:
update transform

IgnaceBleukx committed:
update indicator constraints

IgnaceBleukx committed:
leftover from gurobi

IgnaceBleukx committed:
scip does not support min/max/abs

IgnaceBleukx committed:
update tests and SolverLookup

IgnaceBleukx committed:
add False

IgnaceBleukx committed:
reduce default verbosity

IgnaceBleukx committed:
solveAll

IgnaceBleukx committed:
special cardinality constraints

IgnaceBleukx commented:
Updated the interface somewhat, it seems SCIP does not have the fancy Min/Max/Abs constraints built-in in its API : /

Added the SOS1 and cardinality constraints to the interface too, these constraints are created when decomposing alldiff for example so definately usefull to have them imho. We should also add them to the Gurobi interface

SolveAll is in a weird state for PySCIPOpt: SCIP does allow for retrieving all feasible solutions, but the Python interface does not seem to allow for collecting the solutions which are found... Added the issues to track in the code so we can add the specialized implementation when these are resolved.

All tests are passing so ready for review I think.

Wout4 commented:
I get an error for optimization problems:
```
\cpmpy\solvers\scip.py", line 154, in solve
    self.objective_value_ = scip_objective.getObjVal()
AttributeError: 'pyscipopt.scip.Expr' object has no attribute 'getObjVal'
```

Wout4 committed:
update init

Wout4 commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+        elif scip_status == "infeasible":
+            self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        elif scip_status.endswith("limit"):  # timelimit, nodelimit, ...
+            self.cpm_status.exitstatus = ExitStatus.FEASIBLE
```
This should only be feasible if at least one solution is found, is there a way to check this? Otherwise should be set to unknown to avoid trying to print a solution that doesn't exist later on

Opt-Mucca replied:
To check if a valid solution exists in storage you can call:
```if self.scip_model.getNSols() > 0```

Opt-Mucca commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+            self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+        elif scip_status == "unknown":
+            self.cpm_status.exitstatus = ExitStatus.UNKNOWN
+        # TODO: "userinterrupt" and check if there is a solution?
```
Same as above. Check if there are any solutions in storage and then to obtain the solution call:
```scip_sol = self.scip_model.getBestSol()```

Opt-Mucca commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+                    cpm_var._value = int(solver_val)
+            # set _objective_value
+            if self.has_objective():
+                self.objective_value_ = scip_objective.getObjVal()
```
This should raise an error. The ```getObjVal()``` function can only be called for ```scip_model```, not for the objective expression. So I'd say that you can also remove the line: 
```scip_objective = self.scip_model.getObjective()```

IgnaceBleukx replied:
It did : ) Fixed it and will push it soon

Opt-Mucca commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+            if self.has_objective():
+                self.objective_value_ = scip_objective.getObjVal()
+
+        self.scip_model.freeTransform()  # Mark Turner from SCIP told me you need to do this if you want to support adding additional vars/constraints later...
```
So a tiny summary for this: SCIP transforms the problem that you are actual solving into something it believes is easier to solve (presolve techniques etc). All solutions when found are then transformed back to the original space, because that is obviously what the user modelled and can interpret. Allowing the user to change the model during solving gets messy when balancing these things however, so it's forbidden. What ```self.scip_model.freeTransform()``` does is to remove all information from the transformed space, and just leave the original model. All solutions are kept, and the user can now change the model as they will. The downside is that potentially useful information for speeding up the next optimisation call is thrown out.

hbierlee replied:
documented now

Opt-Mucca commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+                if cpm_var.is_bool(): 
+                    cpm_var._value = solver_val >= 0.5
+                else:
+                    cpm_var._value = int(solver_val)
```
I'm going to assume here that continuous variables just aren't used for CPMpy. If so then this should be fine.

IgnaceBleukx replied:
Jup, indeed we only allow for integer and boolean variables

Opt-Mucca commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+                    sciplhs = self._make_numexpr(lhs)
+                    self.scip_model.addCons(sciplhs <= sciprhs)
+
+            elif cpm_expr.name == '>=':
```
Just a general interest: Due to how the constraints are transformed are you certain that these consraints can't be multiplied by -1 to become cardinality constraints?

IgnaceBleukx replied:
Aha no, actually we are not... I will do a check for it here

IgnaceBleukx replied:
Aha actually no... I'll add a check for it here

Opt-Mucca commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+                    self.scip_model.addCons(sciplhs == sciprhs)
+
+                elif lhs.name == 'mul':
+                    assert len(lhs.args) == 2, "scip only supports multiplication with 2 variables"
```
I'm not sure what is meant here?
SCIP does support more than just quadratic constraints. So you can any order of degree constraint. For example:
```scip_model_addCons(a * b * c * d * e * f * g == sciprhs)``` should work fine

Opt-Mucca commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+                elif lhs.name == 'mul':
+                    assert len(lhs.args) == 2, "scip only supports multiplication with 2 variables"
+                    a, b = self.solver_vars(lhs.args)
+                    #self.scip_model.setParam("NonConvex", 2)  # gurobi leftover
```
This can be removed for SCIP. SCIP allows many more non-linear type formulations than just quadratic.
If you want to give options to SCIP on how to handle quadratic constraints then you can change the existing parameters.

Opt-Mucca commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+                    self.scip_model.addCons(a * b == sciprhs)
+
+                elif lhs.name == 'div':
+                    assert is_num(lhs.args[1]), "scip only supports division by constants"
```
This is also wrong I think..... You can just throw in the constraint regardless if both lhs and rhs are expressions:
```self.scip_model.addCons(a / b == sciprhs)```

Opt-Mucca commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+                raise Exception(f"Unknown linear expression {sub_expr} name")
+
+        # True or False
+        elif isinstance(cpm_expr, BoolVal):
```
Don't really understand what this constraint means, always struggle a bit with the CP language. If you can explain then I can see if there's a more natural MIP approach.

IgnaceBleukx replied:
This is a very uncommon case where a user posts the constraint "False" to the solver, in essence making the solver status trivialy infeasible. 
We run into this case sometimes when our expressions are simplified to "False"

Opt-Mucca commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+            Returns: number of solutions found
+        """
+
+        warnings.warn("Solution enumeration is not implemented in PyScipOPT, defaulting to CPMpy's naive implementation")
```
So I mentioned this is an email. I think it's just that very few MIP practitioners actually need such functionality. To mimic most of this:
- Call optimize for the ```self.scip_model```.
- After solving is completed. Call ```self.scip_model.getSols()```. 
The potential problems with this is that SCIP automatically deletes the "worst" solutions in storage after it has found x many. This is by default 10. See ```limits/maxsol``` and ```limits/maxorigsol```. So to save memory SCIP will at times remove older solutions that it knows are suboptimal because it has clearly better ones. The number of solutions found in total can be retrieved with ```scip_model.getNSolsFound()```, but unless you increase the limits you will not be able to retrieve all of them.

hbierlee replied:
To be handled in https://github.com/CPMpy/cpmpy/issues/932

Opt-Mucca commented:
> I get an error for optimization problems:
> 
> ```
> \cpmpy\solvers\scip.py", line 154, in solve
>     self.objective_value_ = scip_objective.getObjVal()
> AttributeError: 'pyscipopt.scip.Expr' object has no attribute 'getObjVal'
> ```

This should now be addressed in one of my points above

Opt-Mucca commented:
@IgnaceBleukx @tias I haven't run anything, and truthfully haven't tried to understand the larger code base, but I've gone through and given comments that I think should make the interface work.
For a final point: While min / max expressions probably won't work, SCIP can handle the absolute value expressions fairly easily. I'm just not sure where to add them in the current code without some guidance.

IgnaceBleukx committed:
refactor status translation

IgnaceBleukx committed:
minor

IgnaceBleukx committed:
proper objective value retrieving

IgnaceBleukx committed:
remove leftover mul and div restrictions from Gurobi

IgnaceBleukx committed:
remove import

IgnaceBleukx commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,419 @@
+                    lhs = Operator("wsum", [[-w for w in lhs.args[0]], lhs.args[1]])
+                else:
+                    lhs = -lhs
+                self += cond.implies(lhs <= -rhs)
```
Yes, this line sends the new constraint `cond => lhs <= -rhs` through the entire `__add__` pipeline again.
But coming to think of it, we probably pay quite a penalty by doing so as it also sends the constraint through the transformation pipeline...
I'll refactor this somewhat so we avoid unnecessary transformations...

IgnaceBleukx committed:
refactor indicator constraints

tias committed:
Merge branch 'master' into scip2

tias committed:
putting an Agent to work...

tias committed:
Merge remote-tracking branch 'origin/master' into scip2

tias committed:
scip with linear_decompose etc

tias committed:
test solvers: sol limit for all and smaller cumulative

tias committed:
Merge branch 'master' into scip2

tias committed:
update tests

IgnaceBleukx committed:
add scip requirements and ci

IgnaceBleukx committed:
fix error raising in supported check

IgnaceBleukx committed:
pyscipopt is typed

IgnaceBleukx committed:
re-add pyscopopt to mypy.ini

tias committed:
Merge remote-tracking branch 'origin/master' into scip2

tias committed:
Merge remote-tracking branch 'origin/master' into scip2

tias committed:
scip: fix imports

tias committed:
scip: div fix

tias committed:
doc updates

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+        Constructor of the native solver object
+
+        Arguments:
+        - cpm_model: a CPMpy Model()
```
missing arg description (although I'd say if the doc is the same as the superclass, not including it will inherit that doc)

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+        self.scip_model = scip.Model("From CPMpy")
+        self.scip_model.setParam("display/verblevel", 0) # remove solver logs from output
+        # initialise everything else and post the constraints/objective
+        # it is sufficient to implement __add__() and minimize/maximize() below
```
superfluous comment(s)?

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+            Call the SCIP solver
+
+            Arguments:
+            - time_limit:  maximum solve time in seconds (float, optional). Persists across solve() calls until overridden.
```
is this persistence the same behaviour as in other solvers? Any way to make this uniform? I also don't really understand what is meant with it (with time_limit=10, if the first call is 6 seconds, will the next call have a limit of 4 seconds, or 10 again, or will setting time_limit=None overwrite it?)

tias replied:
I would expect it means that it stays 10 seconds for subsequent calls, just doc'ing it is OK for now I think

hbierlee replied:
this is a bit subtle, note to self, I'll add a solver interface test for this.

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+            - kwargs:      any keyword argument, sets parameters of solver object (e.g. "limits/time" overrides time_limit).
+
+            Arguments that correspond to solver parameters (passed via setParams):
+            Examples: limits/nodes=1000, limits/solutions=1, parallel/maxnthreads=4,
```
how does one pass an arg "limits/nodes" as kwarg?

hbierlee replied:
I suppose as `**dict` may be worth indicating

tias replied:
idd

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+            self.scip_model.setParam("limits/time", float(time_limit))
+
+        if solution_callback is not None:
+            raise NotSupportedError("SCIP: solution callback not (yet?) implemented")
```
Meaning SCIP does not support it or CPMpy? If latter, should it be in this PR?

tias replied:
it means CPMpy has not implemented it; a PR choses its own scope so it can remain unimplemented

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+        self.cpm_status = SolverStatus(self.name)
+        self.cpm_status.runtime = self.scip_model.getSolvingTime()
+
+        unknown_stati = {"unknown", "userinterrupt", "unbounded", "inforunbd", "terminate"}
```
stati?

hbierlee replied:
apparently statuses is more correct/widely used

hbierlee replied:
not reused, could just be inline in 161

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+            self.cpm_status.exitstatus = ExitStatus.UNSATISFIABLE
+        else:
+            n_sols = self.scip_model.getSols()
+            n_sols = len(n_sols) if isinstance(n_sols, list) else n_sols
```
on what does this dpeend? I'd expect a comment on why this is necessary

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+        else:
+            n_sols = self.scip_model.getSols()
+            n_sols = len(n_sols) if isinstance(n_sols, list) else n_sols
+            if scip_status.endswith("limit"):  # timelimit, nodelimit, etc.
```
I'd list the options here (although painful it will lead to less unexpected behavior)

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                self.cpm_status.exitstatus = ExitStatus.FEASIBLE if n_sols > 0 else ExitStatus.UNKNOWN
+            elif n_sols > 0:  # at least one feasible solution found, not proven optimal
+                self.cpm_status.exitstatus = ExitStatus.FEASIBLE
+            elif scip_status in unknown_stati:
```
I feel like this whole section can be just matching on all these statuses, then checking n_sols, no need for all these casees

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                    continue  # variable never posted (e.g. unused); skip to avoid addVar after solve
+                scip_var = self._varmap[cpm_var]
+                if best_sol is not None:
+                    solver_val = self.scip_model.getSolVal(best_sol, scip_var)
```
might want to check if we can access multiple vars at once?

hbierlee replied:
doesn't seem to be possible

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+            best_sol = self.scip_model.getBestSol()
+            for cpm_var in self.user_vars:
+                if cpm_var not in self._varmap:
+                    continue  # variable never posted (e.g. unused); skip to avoid addVar after solve
```
this is dangerous as then that user variable is unassigned. Since we add all model variables, this should be an assert

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+            for cpm_var in self.user_vars:
+                if cpm_var not in self._varmap:
+                    continue  # variable never posted (e.g. unused); skip to avoid addVar after solve
+                scip_var = self._varmap[cpm_var]
```
why not via self.solver_var() ?

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                if cpm_var.is_bool():
+                    cpm_var._value = solver_val >= 0.5
+                else:
+                    cpm_var._value = int(solver_val)
```
this one should also be round  like said in #884

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+        # What self.scip_model.freeTransform() does is to remove all information from the 
+        # transformed space, and just leave the original model. All solutions are kept, 
+        # and the user can now change the model as they will. The downside is that potentially 
+        # useful information for speeding up the next optimisation call is thrown out.
```
I think a) this comment should be just summarized, also b) if freeTransform is necessary before you incrementally solve anyway (unless you re-solve the exact same problem) then this downside seems unavoidable? Is there any useful way to incrementally solve without calling freeTransform in-between calls?

tias replied:
This is a quote from SCIP/Mark directly, to avoid misinterpreting it is can be good to keep it as is.

we do it because we want to support incrementality; if you don't need incrementality you don't need it...
its to document the motivation; e.g. somebody at some point might consider adding a flag to see if solve has already been called before and only then do this. documenting is future-proofing

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+        try:
+            obj = self.scip_model.getObjective()
+            return obj is not None and (hasattr(obj, 'terms') and len(obj.terms) != 0)
+        except (AttributeError, TypeError):
```
unnecessary caught exceptions?

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+    def has_objective(self):
+        try:
+            obj = self.scip_model.getObjective()
+            return obj is not None and (hasattr(obj, 'terms') and len(obj.terms) != 0)
```
could be `obj is not None and getattr(obj, 'terms', False)`

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+            Turns a numeric CPMpy 'flat' expression into a solver-specific
+            numeric expression
+
+            Used especially to post an expression as objective function
```
it's used equally when adding constraints ; superfluous doc imo

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+
+        # negated bool (e.g. in objective): 1 - bv
+        if isinstance(cpm_expr, NegBoolView):
+            return 1 - self.solver_var(cpm_expr._bv)
```
needed? in gurobi, we remove all NegBoolViews

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+        # GlobalFunction: abs, mul (PySCIPOpt supports these in constraints/objective)
+        if isinstance(cpm_expr, GlobalFunction):
+            if cpm_expr.name == "abs":
+                return abs(self._make_numexpr(cpm_expr.args[0]))
```
but they will still be needlessly flattened right? so this might be another usecase for #912

hbierlee replied:
this can just be updated in #912

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+        return cpm_cons
+
+    def add(self, cpm_expr_orig):
+        """
```
I think this entire doc is duplicated so can be inherited

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+        # Ensure every user var has a solver variable (so we get values after solve even if
+        # the constraint was simplified away and the var never appears in transformed constraints)
+        for cpm_var in self.user_vars:
+            if not is_num(cpm_var):
```
is_num already checked in solver_var

I think just one-liner suffices with : self.solver_Vars(self.user_vars)

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                if cpm_expr.name == '!=':
+                    # sum(x) != k  =>  (lhs <= k-1) or (lhs >= k+1); use SCIP's native disjunction (no extra binary)
+                    sciplhs = self._make_numexpr(lhs)
+                    self.scip_model.addConsDisjunction([
```
this tells me we should leave in just.. disjunctions?

hbierlee replied:
I believe this would require partial flattening e.g. what #912 is trying to accomplish

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                        sciplhs <= sciprhs - 1,
+                        sciplhs >= sciprhs + 1
+                    ])
+                    continue
```
just using if .. elif .. else would avoid continue's I'd say

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                    continue
+
+                if cpm_expr.name == '<=':
+                    if (lhs_is_operator and lhs.name == "sum" and all(a.is_bool() and not isinstance(a, NegBoolView) for a in lhs.args)):
```
we should consider adding SOS constraints for gurobi as well! Just making this comment to be converted to issue if agreed

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                assert isinstance(sub_expr, Comparison), "Implication must have linear constraints on right hand side"
+
+                lhs, rhs = sub_expr.args
+                lhs_is_globalfunc = isinstance(lhs, GlobalFunction)
```
these variables can be avoided at no cost I think

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                lhs_is_globalfunc = isinstance(lhs, GlobalFunction)
+                lhs_is_operator = isinstance(lhs, Operator)
+                if lhs_is_globalfunc and lhs.name == "abs":
+                    (arg,) = lhs.args
```
the neccessity here took some time to understand, maybe add a comment. So for reified abs, we make an ad-hoc transformation here?

hbierlee replied:
the others do work reified?

hbierlee replied:
actually, no global is left in reified form, so this code turned out to be dead as revealed by code coverage

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                        lhs = Operator("wsum", [[-w for w in lhs.args[0]], lhs.args[1]])
+                    else:
+                        lhs = Operator("wsum", [[-1], [lhs]])
+                    sub_expr = lhs <= -rhs
```
why no indicator constraint? Seems supported https://pyscipopt.readthedocs.io/en/latest/tutorials/constypes.html

hbierlee replied:
ok it's used below. I'm a bit lost as to what is exactly posted here?

tias replied:
it seems to ensure that lhs is a wsum (the one at 430 is needlessly recreating a wsum)

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                        lhs = Operator("wsum", [[-1], [lhs]])
+                    sub_expr = lhs <= -rhs
+
+                if sub_expr.name == "<=":
```
why is the <= case so different from the >= case?

tias replied:
silly agents...

hbierlee replied:
interestingly because the SCIP docs actually said only `<=` was allowed in indicators! Yet this is actually not true. I reported this incorrect doc https://github.com/scipopt/PySCIPOpt/issues/1207

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+            elif isinstance(cpm_expr, BoolVal):
+                if cpm_expr.args[0] is False:
+                    bv = self.solver_var(boolvar())
+                    self.scip_model.addCons(bv <= -1)
```
no e.g. 0 <= -1  possible?

hbierlee replied:
nop

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                    if not args:
+                        # only false constants: xor() is False -> no solution
+                        if not hasattr(self, "_scip_infeasible_aux"):
+                            self._scip_infeasible_aux = self.scip_model.addVar(vtype='B', lb=0, ub=0, name='_false')
```
just can add(BoolVal(False)), avoiding a bit of code here

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                        continue
+                    if any(is_true_cst(a) for a in args):
+                        if not hasattr(self, "_scip_true_var"):
+                            self._scip_true_var = self.scip_model.addVar(vtype='B', lb=1, ub=1, name='true')
```
_true

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+                            self._scip_infeasible_aux = self.scip_model.addVar(vtype='B', lb=0, ub=0, name='_false')
+                            self.scip_model.addCons(self._scip_infeasible_aux >= 1)
+                        continue
+                    if any(is_true_cst(a) for a in args):
```
needless if check? Can't we just have line 471 to replace constants and variables by scip constants and variables?

hbierlee replied:
this whole block: `[self._scip_true_var if is_true_cst(a) else self.solver_var(a) for a in args if not is_false_cst(a)]` seems more readable and faster

hbierlee commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,514 @@
+        """
+            Issues to track for future reference:
+                https://github.com/scipopt/PySCIPOpt/issues/549 and
+                https://github.com/scipopt/PySCIPOpt/issues/248
```
both these issues are now closed so perhaps we can do it natively now?

tias replied:
No need to extend the PR...

hbierlee replied:
yup, I'll add a new issue though:

With https://github.com/scipopt/PySCIPOpt/issues/549 and https://github.com/scipopt/PySCIPOpt/issues/248 now resolved, we can enumerate solutions for SCIP natively using `getSols` rather than with the (presumably) slower CPMpy generic implementation.

hbierlee reviewed (changes requested):
@tias I've reviewed the PR, I have many small comments and a few larger ones. Is the idea now that I'll resolve them and nominate the PR for review?

tias commented:
> @tias I've reviewed the PR, I have many small comments and a few larger ones. Is the idea now that I'll resolve them and nominate the PR for review?

Normally the one that created the PR would be the one fixing the comments.

In this case, the PR was written partly by Ignace, me and more recently an agent. Hence possible inconsistencies and overcomplications.

So, if you would be willing to go over it and clean it up (I commented on the more open questions), then that would be much appreciated!

hbierlee commented:
yep perfect that's what I had in mind.

hbierlee committed:
Process some of the smaller comments

hbierlee committed:
Merge branch 'master' into scip2

hbierlee committed:
Some additions to test_solverinterface to be split out

hbierlee committed:
Process remaining review comments

hbierlee committed:
Remove unused reified abs case

hbierlee committed:
Actually use the SCIP != case

hbierlee committed:
Remove != (or rather Disjunction) support

hbierlee committed:
Remove dead `sub` support

hbierlee committed:
Refactor _make_numexpr

hbierlee committed:
Merge branch 'master' into scip2

hbierlee committed:
Remove changes to `test_solverinterface`

hbierlee committed:
Polish

hbierlee committed:
Remove duplicated docstrings

hbierlee committed:
Polish PR

hbierlee committed:
Polish

tias committed:
typo/consistency

tias commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,387 @@
+        return has_sol
+
+
+    def solver_var(self, cpm_var):
```
do we do this function the same elsewhere?

its not written with 'fast path' in mind... an alterantive would be
```
solver_var = self._varmap.get(cpm_var, None)

if solver_var = None:
  if is_num...
  elif is _BoolVarImpl:
    if NegBool: raise
    else:
       solver_var = ...
  elif _IntVarImple:
  self._varmap[cpm_var] = solver_var

return solver_var
```

if other solvers do similar like we do here, we should fix this in a separate PR

hbierlee replied:
It does seem indeed more or less copy-paste from gurobi, but it's not the fast way. I agree it's pretty sub-optimal, coverage revealed the is_num is never run, similarly NegBoolView should never run as long as the transforms are correct.

https://github.com/CPMpy/cpmpy/blob/626bfa80c6507bf2ada8aa5d665766ceb031af75/cpmpy/solvers/gurobi.py#L252-L276

I'll convert this comment to issue

tias committed:
tweak

tias commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,387 @@
+
+    def has_objective(self):
+        obj = self.scip_model.getObjective()
+        return obj is not None and getattr(obj, 'terms', False)  # obj could be `Expr({})`
```
I don't get this one; if it is Expr, is it not valid?

ah, or specifically if it is an Expr of empty list? Can you tighten your check?

tias commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,387 @@
+
+        if is_num(cpm_expr):
+            return cpm_expr
+        elif isinstance(cpm_expr, NegBoolView):  # negated bool (e.g. in objective): 1 - bv
```
check can go, it would match the next case, where it would be caught in solver_var

tias commented on cpmpy/solvers/scip.py:
```diff
@@ -0,0 +1,387 @@
+        elif isinstance(cpm_expr, DirectConstraint):
+            cpm_expr.callSolver(self, self.scip_model)
+
+        elif isinstance(cpm_expr, GlobalConstraint):
```
I think it is a tiny-tiny bit neater for this one to appear before the isisnstance BoolVal

tias reviewed (commented):
very nice and clean, made some minor tweaks and minor comments

(also a less minor comment that is just triggered by the review, need not be for the PR)

tias committed:
too quick...

hbierlee committed:
More explicit has_objective check

- Now similar to Gurobi
- Also, `getObjective` should always return Expr according to scip docs

hbierlee committed:
Rm superfluous test

hbierlee committed:
Move GlobalConstraint case

hbierlee committed:
Merge branch 'master' into scip2

hbierlee committed:
Fix syntax

tias reviewed (commented):
I approve very much, but can't formally approve because I started the PR indeed : )

But its your earlier review that is blocking it, so please approve if you approve and then merge ; )
